cmd/swarmctl/service/flagparser/resource.go

```diff
@@ -3,90 +3,122 @@ package flagparser
 import (
 	"fmt"
 	"math/big"
+	"strconv"
+	"strings"
 
 	"github.com/docker/go-units"
 	"github.com/docker/swarmkit/api"
 	"github.com/spf13/pflag"
 )
 
-func parseResourceCPU(flags *pflag.FlagSet, resources *api.Resources, name string) error {
-	cpu, err := flags.GetString(name)
+var (
+	// string name of pre-defined resources
+	cpu         = api.CPUs.String()
+	memory      = api.Memory.String()
+	nanoCPUs    = api.NanoCPUs.String()
+	memoryBytes = api.MemoryBytes.String()
+)
+
+// ParseUserDefinedResources parses user-defined resources from string to api.Resources
+func ParseUserDefinedResources(flags *pflag.FlagSet, resources *api.Resources, name string, isLimit bool) error {
+	resourcesStr, err := flags.GetStringSlice(name)
 	if err != nil {
 		return err
 	}
 
-	nanoCPUs, ok := new(big.Rat).SetString(cpu)
-	if !ok {
-		return fmt.Errorf("invalid cpu: %s", cpu)
-	}
-	cpuRat := new(big.Rat).Mul(nanoCPUs, big.NewRat(1e9, 1))
-	if !cpuRat.IsInt() {
-		return fmt.Errorf("CPU value cannot have more than 9 decimal places: %s", cpu)
+	if len(resourcesStr) == 0 {
+		return nil
 	}
-	resources.NanoCPUs = cpuRat.Num().Int64()
-	return nil
-}
 
-func parseResourceMemory(flags *pflag.FlagSet, resources *api.Resources, name string) error {
-	memory, err := flags.GetString(name)
-	if err != nil {
-		return err
-	}
+	// convert all ResourcesName to lower case and check
+	// 0. if resource type is scalar
+	// 1. if re-define resources
+	// 2. validate pre-defined resources
+	// 3. validate if other user-defined resources value is number
+	definedResources := make(map[string]bool, len(resourcesStr))
+	for _, resource := range resourcesStr {
+		parts := strings.Split(resource, "=")
+		if len(parts) != 2 {
+			return fmt.Errorf("error format in resources configuration, please define as [name]=[value]")
+		}
 
-	bytes, err := units.RAMInBytes(memory)
-	if err != nil {
-		return err
+		newName := strings.ToLower(parts[0])
```

Comment:
@dongluochen  I think part[0] is already trimed, because `--reservation cpu   =0.5`  will throw an error by bash. And the leading spaces are removed by `pflag` automatically.